File: backend/src/translation_validation.rs
use crate::{
    errors::{ApiError, AppResult},
    util::{optional_trimmed, required_non_empty},
};

pub const MAX_TRANSLATION_KEY_LEN: usize = 500;
pub const MAX_TRANSLATION_VALUE_LEN: usize = 10_000;
pub const MAX_TRANSLATION_DESCRIPTION_LEN: usize = 2_000;
pub const MAX_TRANSLATION_IMPORT_ENTRIES: usize = 5_000;

pub fn validate_translation_key<'a>(value: &'a str, namespace: &str) -> AppResult<&'a str> {
    let key = required_non_empty(value, "Translation key cannot be empty.")?;

    if key.chars().count() > MAX_TRANSLATION_KEY_LEN {
        return Err(ApiError::validation(format!(
            "Translation key {key:?} must be at most {MAX_TRANSLATION_KEY_LEN} characters."
        )));
    }

    if key
        .chars()
        .any(|character| character.is_control() || matches!(character, ':' | '{' | '}'))
    {
        return Err(ApiError::validation(format!(
            "Translation key {key:?} contains unsupported characters. Colons, braces, and control characters are not allowed."
        )));
    }

    let namespace = namespace.trim();
    if !namespace.is_empty() && key.starts_with(&format!("{namespace}.")) {
        return Err(ApiError::validation(format!(
            "Translation key {key:?} must be local to namespace {namespace:?} and must not include the namespace prefix."
        )));
    }

    Ok(key)
}

pub fn validate_translation_value<'a>(value: &'a str, key: Option<&str>) -> AppResult<&'a str> {
    let field = key
        .map(|key| format!("Translation value for key {key:?}"))
        .unwrap_or_else(|| "Translation value".to_owned());
    let value = value.trim();
    if value.is_empty() {
        return Err(ApiError::validation(format!("{field} cannot be empty.")));
    }

    if value.chars().count() > MAX_TRANSLATION_VALUE_LEN {
        return Err(ApiError::validation(format!(
            "{field} must be at most {MAX_TRANSLATION_VALUE_LEN} characters."
        )));
    }

    Ok(value)
}

pub fn validate_translation_description(value: Option<&str>) -> AppResult<Option<&str>> {
    let description = optional_trimmed(value);
    if let Some(description) = description
        && description.chars().count() > MAX_TRANSLATION_DESCRIPTION_LEN
    {
        return Err(ApiError::validation(format!(
            "Description must be at most {MAX_TRANSLATION_DESCRIPTION_LEN} characters."
        )));
    }

    Ok(description)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn key_validation_accepts_local_unicode_keys_and_rejects_unsupported_formats() {
        assert_eq!(
            validate_translation_key("  checkout.кнопка-save_1  ", "common").expect("valid key"),
            "checkout.кнопка-save_1"
        );

        for invalid in [
            "common.button.save",
            "button:save",
            "button{save}",
            "button\nsave",
        ] {
            assert!(
                validate_translation_key(invalid, "common").is_err(),
                "{invalid:?}"
            );
        }
    }

    #[test]
    fn translation_limits_count_unicode_characters_and_normalize_description() {
        assert!(validate_translation_key(&"я".repeat(MAX_TRANSLATION_KEY_LEN), "common").is_ok());
        assert!(
            validate_translation_key(&"я".repeat(MAX_TRANSLATION_KEY_LEN + 1), "common").is_err()
        );
        assert!(
            validate_translation_value(&"я".repeat(MAX_TRANSLATION_VALUE_LEN + 1), None).is_err()
        );
        assert_eq!(
            validate_translation_description(Some("   ")).expect("empty description"),
            None
        );
    }
}

File: frontend/src/lib/translationValidation.ts
export const MAX_TRANSLATION_KEY_LENGTH = 500;
export const MAX_TRANSLATION_VALUE_LENGTH = 10_000;
export const MAX_TRANSLATION_IMPORT_ENTRIES = 5_000;

export function validateTranslationKey(key: string, namespace: string): string | null {
  const normalizedKey = key.trim();
  if (!normalizedKey) {
    return "Every translation key must be a non-empty string.";
  }

  if (characterCount(normalizedKey) > MAX_TRANSLATION_KEY_LENGTH) {
    return "Translation key " + JSON.stringify(normalizedKey) + " must be at most " +
      MAX_TRANSLATION_KEY_LENGTH + " characters.";
  }

  if (/[:{}\u0000-\u001f\u007f-\u009f]/u.test(normalizedKey)) {
    return "Translation key " + JSON.stringify(normalizedKey) +
      " contains unsupported characters. Colons, braces, and control characters are not allowed.";
  }

  const normalizedNamespace = namespace.trim();
  if (normalizedNamespace && normalizedKey.startsWith(normalizedNamespace + ".")) {
    return "Translation key " + JSON.stringify(normalizedKey) + " must be local to namespace " +
      JSON.stringify(normalizedNamespace) + " and must not include the namespace prefix.";
  }

  return null;
}

export function validateTranslationValue(value: string, key?: string): string | null {
  const normalizedValue = value.trim();
  const field = key ? "The value for " + JSON.stringify(key) : "Translation value";
  if (!normalizedValue) {
    return field + " cannot be empty.";
  }

  if (characterCount(normalizedValue) > MAX_TRANSLATION_VALUE_LENGTH) {
    return field + " must be at most " + MAX_TRANSLATION_VALUE_LENGTH + " characters.";
  }

  return null;
}

function characterCount(value: string) {
  return Array.from(value).length;
}
